### ISSUE TITLE
[Question] How to use launchPersistentContext with Playwright Test

### ISSUE THREAD

**Issue by badsyntax** [microsoft/playwright#7447]
Hi. I'm attempting to run lighthouse on browser instances managed by playwright. 

Lighthouse will launch a new window/session and the auth state is not persisted between windows. 

To allow persistent auth I can use `launchPersistentContext`. This works fine outside of playwright test:

```js
const userDataDir = '/path/to/data/dir';

const context = await chromium.launchPersistentContext(userDataDir, {
  args: ['--remote-debugging-port=9222'],
});
const page = await context.newPage();
await page.goto('http://localhost:3000/');

// Perform login steps here which will save to cookie or localstorage

// When lighthouse opens a new page the storage will be pesisted meaning the new page will have the same session (either via cookie or localstorage)
await lighthouseImpl({
  page: page,
  port: 9222,
});
```

How can I use `launchPersistentContext` with the Playwright Test Runner? I've attempted to override the browser context using fixtures, but I get the following error:

```
Fixture "context" has already been registered as a { scope: 'test' } fixture.
```

I am able to override the `browser` fixture but I don't have access to a `browser` instance when using `launchPersistentContext`. Looking for some tips or pointers, thanks in advance!

**Comment by pavelfeldman** [comment]
There are several ways of addressing your issue. Could you give us a hint on how you use lighthouse? Do you want one of your tests to simply generate a lighthouse report? Do you do it more than once? Do you perform actions after report is generated?

**Comment by badsyntax** [comment]
Hi @pavelfeldman

> Could you give us a hint on how you use lighthouse?

I'm using a [convenience library](https://github.com/abhinaba-ghosh/playwright-lighthouse#readme) to run lighthouse. The library doesn't do much, it essentially calls lighthouse like so:

```ts
import lighthouse from 'lighthouse';

const results = await lighthouse(
  url,
  { disableStorageReset: true, port: port },
  config
);
````

> Do you want one of your tests to simply generate a lighthouse report?

To both generate a report and to fail the tests if the thresholds don't match. 

> Do you do it more than once?

Yes, I'll run lighthouse in different tests. I achieve this without breaking the parallelism by overriding the browser fixture to pass in a random unused remote debugging port which lighthouse uses to connect to the correct browser instance:

```ts
import { chromium } from 'playwright';
import type { Browser } from 'playwright';
import { test as base } from '@playwright/test';
import getPort from 'get-port';

export const lighthouseTest = base.extend<
  {},
  { port: number; browser: Browser }
>({
  port: [
    async ({}, use) => {
      // We need to assign a unique port for each playwright worker to allow
      // lighthouse tests to run in parallel
      const port = await getPort();
      await use(port);
    },
    { scope: 'worker' },
  ],

  // @ts-ignore
  browser: [
    async ({ port }, use) => {
      const browser = await chromium.launch({
        args: [`--remote-debugging-port=${port}`, '--lang=en-GB'],
      });
      await use(browser);
    },
    { scope: 'worker' },
  ],
});
```

I'll run lighthouse like so:

```ts
import { rootUrl } from '../../constants';
import { lighthouseTest } from '../../e2eUtil';
import { playAudit } from 'playwright-lighthouse';

lighthouseTest.describe('Login screen lighthouse', () => {
  lighthouseTest('should pass lighthouse tests', async ({ page, port }) => {
    await page.goto(`${rootUrl}/login`);
    await page.waitForSelector('[data-testid="login-button"]');
    await playAudit({
      page: page,
      reports: {
        formats: {
          html: true,
          json: true,
          csv: false,
        },
        name: `lhr-${new Date().getTime()}`,
        directory: './.lighthouse',
      },
      thresholds: {
        performance: 70,
        accessibility: 95,
        pwa: 75,
        'best-practices': 95,
      },
      opts: {
        logLevel: 'error',
      },
      port,
    });
  });
});
```

> Do you perform actions after report is generated?

After tests are run the lighthouse reports are uploaded to azure blob storage and a PR comment is created to add links to the lighthouse reports.

Regarding running lighthouse behind authenticated pages, their recommended approach is to [use puppeteer](https://github.com/GoogleChrome/lighthouse/blob/master/docs/recipes/auth/README.md). [Another guide here](https://github.com/GoogleChrome/lighthouse-ci/blob/main/docs/configuration.md#puppeteerscript). The concept being you use puppeteer to login the user in (to create the user session), then close the page, the pass off to lighthouse which will open a new page. This works because (i believe) puppeteer doesn't open in incognito. This approach doesn't work for playwright because it open in an incognito and no auth state is shared with new windows.

**Comment by badsyntax** [comment]
Another alternative approach is to "manually" inject auth (in my case set localstorage keys) in the lighthouse page. To do this we need an API to know when a new page is opened from a browser. Pupeteer provides [`browser.on('targetcreated')`](https://pptr.dev/#?product=Puppeteer&version=v10.1.0&show=api-event-targetcreated). Is there something similar for playwright?

**Comment by badsyntax** [comment]
@pavelfeldman got any advice? for the meantime i will have to run my lighthouse tests outside of the playwright test runner. 

For my usecase it would be ideal if:

- We can somehow use launchPersistentContext within playwright test runner
- If playwright exposed an API to determine when new pages/contexts are opened, similar to [`browser.on('targetcreated')`](https://pptr.dev/#?product=Puppeteer&version=v10.1.0&show=api-event-targetcreated) 
  -  I'd like this API to allow me to perform setup tasks on the page opened by lighthouse (like mocking network requests)

**Comment by dgozman** [comment]
@badsyntax Let me answer a few questions here

> If playwright exposed an API to determine when new pages are created, similar to browser.on('targetcreated')
> I'd like this API to allow me to perform setup tasks on the page opened by lighthouse (like mocking network requests)

You can setup [routing on the context](https://playwright.dev/docs/api/class-browsercontext#browser-context-route) - this way it will affect any pages open in this context. Overall, `BrowserContext` should expose all the APIs you need to work with multiple pages out of the box.

> We can somehow use launchPersistentContext within playwright test runner

You can do everything manually if you'd want. E.g. write a helper that will:
- Call `chromium.launchPersistentContext` with a port, open a page there, perform actions (to authenticate?).
- Call into Lighthouse
- Close the context
Then use this helper from any of your tests. For this you don't need any of the default Playwright Test fixtures, you can just use any Playwright APIs that work for you.

Meanwhile, we'll collect more usecases for persistent context to make a call where we want some kind of built-in API for this in Playwright Test and how it should look like.

**Comment by badsyntax** [comment]
@dgozman thanks. wasn't aware of routing on the context, that's super helpful. thanks for these suggestions, i'll try them out when i have some time and give some feedback, but i reckon this approach should work.

**Comment by erix-gdrx** [comment]
👀  following. We've got the same issue. Want to throw lighthouse into our E2E tests to capture Web Vitials as part of our CI runs. But lighthouse can't get into the authenticated pages even though Playwright already logged in on the test.

**Comment by badsyntax** [comment]
Ok i've got this to work with the Test runner.

Here's my `e2eUtil.ts` file with playwright fixtures:

```ts
import os from 'os';
import getPort from 'get-port';
import { playAudit } from 'playwright-lighthouse';
import { test as base } from '@playwright/test';
import { BrowserContext, chromium, Page } from 'playwright';
// @ts-ignore
import lighthouseMobileConfig from 'lighthouse/lighthouse-core/config/lr-mobile-config.js';
import { rootUrl } from './constants';

// auth-state.json is saved during globalSetup
import authState from './fixtures/auth-state.json';

type LocalStorage = Array<{ name: string; value: string }>;

type AuthState = {
  origins: Array<{
    origin: string;
    localStorage: Array<{
      name: string;
      value: string;
    }>;
  }>;
};

export const lighthouseTest = base.extend<
  {
    lighthouseAuthenticatedPage: Page;
    lighthousePage: Page;
    persistentContext: BrowserContext;
  },
  {
    port: number;
    storage: LocalStorage;
  }
>({
  // We need to assign a unique port for each playwright worker to allow
  // lighthouse tests to run in parallel
  port: [
    async ({}, use) => {
      const port = await getPort();
      await use(port);
    },
    { scope: 'worker' },
  ],

  // As there's currently no way to insert auth-state into a PersistentContext we
  // need to explicity load and insert the auth storage. In our case we only care
  // about localStorage for auth.
  storage: [
    async ({}, use) => {
      const localhostStorage = (authState as AuthState).origins.find(
        ({ origin }) => {
          return origin === new URL(rootUrl).origin;
        }
      );
      if (!localhostStorage) {
        throw new Error(`No storage state found in ${authStatePath}`);
      }
      if (!localhostStorage.localStorage.length) {
        throw new Error(`No localStorage state found in ${authStatePath}`);
      }

      await use(localhostStorage.localStorage);
    },
    { scope: 'worker' },
  ],

  // As lighthouse opens a new page, and as playwright does not by default allow
  // shared contexts, we need to explicitly create a persistent context to
  // allow lighthouse to run behind authenticated routes.
  persistentContext: [
    async ({ port }, use) => {
      const userDataDir = os.tmpdir();
      const context = await chromium.launchPersistentContext(userDataDir, {
        args: [`--remote-debugging-port=${port}`],
      });

      await use(context);
    },
    { scope: 'test' },
  ],

  lighthousePage: [
    async ({ persistentContext: context }, use) => {
      const page = await context.newPage();
      await use(page);
    },
    { scope: 'test' },
  ],

  lighthouseAuthenticatedPage: [
    async (
      { persistentContext: context, lighthousePage: page, storage },
      use
    ) => {
      // The following uses context.route() to mock some auth related requests (token validation)
      await mockB2CLogin(context);

      await page.goto(rootUrl);
      await insertAuthStorage(page, storage);
      await use(page);
    },
    { scope: 'test' },
  ],
});

type LighthouseThresholds = {
  performance?: number;
  accessibility?: number;
  pwa?: number;
  'best-practices'?: number;
};

export async function lighthouseAudit(
  page: Page,
  port: number,
  thresholds?: LighthouseThresholds,
  config = lighthouseMobileConfig
): Promise<void> {
  await playAudit({
    page,
    config,
    port,
    reports: {
      formats: {
        html: true,
        json: true,
        csv: false,
      },
      name: `lhr-${new Date().getTime()}`,
      directory: './.lighthouse',
    },
    thresholds: {
      performance: 70,
      accessibility: 95,
      pwa: 50,
      'best-practices': 95,
      ...thresholds,
    },
    opts: {
      logLevel: 'error',
    },
  });
}
```

Running lighthouse on a public route:

```ts
import { rootUrl } from '../../constants';
import { lighthouseAudit, lighthouseTest } from '../../e2eUtil';

lighthouseTest.describe('Login screen lighthouse', () => {
  lighthouseTest(
    'should pass lighthouse tests',
    async ({ lighthousePage: page, port }) => {
      await page.goto(rootUrl);
      await page.waitForSelector('[data-testid="login-button"]');
      await lighthouseAudit(page, port);
    }
  );
});
```

Running lighthouse on an authenticated route:

```ts
import { rootUrl } from '../../constants';
import { lighthouseAudit, lighthouseTest } from '../../e2eUtil';

lighthouseTest.describe('Dashboard screen lighthouse', () => {
  lighthouseTest(
    'should pass lighthouse tests',
    async ({ port, lighthouseAuthenticatedPage: page }) => {
      await page.goto(rootUrl);
      await page.innerText('[data-testid="app-bootstrapped"]');
      await lighthouseAudit(page, port);
    }
  );
});
```

I'd appreciate a quick review to validate I'm not doing anything wrong with this approach.

@dgozman It would be convenient if `chromium.launchPersistentContext` provided a way to populate auth storage similar to [browser.newContext({ storageState: 'state.json' })](https://playwright.dev/docs/api/class-browser#browser-new-context)

**Comment by dgozman** [comment]
@badsyntax This seems legit. A few comments:

- You are not closing the persistentContext, and you probably want to:
  ```js
    ...
    await use(context);
    await context.close();
  ```

- You can override page/context fixtures if you'd like to avoid renaming like this: `{ lighthouseAuthenticatedPage: page }`.
  ```js
  context: [
    async ({ port }, use) => {
      const userDataDir = os.tmpdir();
      const context = await chromium.launchPersistentContext(userDataDir, {
        args: [`--remote-debugging-port=${port}`],
      });
      await use(context);
      await context.close();
    },
    { scope: 'test' },
  ],
  ```
  If you do this, default `page` should pick up your persistent `context` right away.

**Comment by dgozman** [comment]
> It would be convenient if chromium.launchPersistentContext provided a way to populate auth storage similar to browser.newContext({ storageState: 'state.json' })

Please file a separate issue for this, so we can prioritize accordingly.

**Comment by badsyntax** [comment]
Awesome thank you @dgozman! I've made the change you suggested and `page` uses the persistent `context` as you described. This approach is a bit neater. 

FYI there's a TypeScript type conflict when overriding the `context` fixture and it's not obvious why. It's not a huge issue as i'm using `@ts-ignore` to suppress the error.

<details><summary>Error here</summary>

```console
Type '[({ port }: { authenticatedPage: Page; context: BrowserContext; } & { port: number; storage: LocalStorage; } & PlaywrightTestArgs & PlaywrightTestOptions & PlaywrightWorkerArgs & PlaywrightWorkerOptions, use: (r: BrowserContext) => Promise<...>) => Promise<...>, { ...; }]' is not assignable to type '(TestFixtureValue<BrowserContext, { authenticatedPage: Page; context: BrowserContext; } & { port: number; storage: LocalStorage; } & PlaywrightTestArgs & PlaywrightTestOptions & PlaywrightWorkerArgs & PlaywrightWorkerOptions> & (TestFixtureValue<...> | [...])) | undefined'.
  Type '[({ port }: { authenticatedPage: Page; context: BrowserContext; } & { port: number; storage: LocalStorage; } & PlaywrightTestArgs & PlaywrightTestOptions & PlaywrightWorkerArgs & PlaywrightWorkerOptions, use: (r: BrowserContext) => Promise<...>) => Promise<...>, { ...; }]' is not assignable to type 'TestFixture<BrowserContext, { authenticatedPage: Page; context: BrowserContext; } & { port: number; storage: LocalStorage; } & PlaywrightTestArgs & PlaywrightTestOptions & PlaywrightWorkerArgs & PlaywrightWorkerOptions> & [...]'.
    Type '[({ port }: { authenticatedPage: Page; context: BrowserContext; } & { port: number; storage: LocalStorage; } & PlaywrightTestArgs & PlaywrightTestOptions & PlaywrightWorkerArgs & PlaywrightWorkerOptions, use: (r: BrowserContext) => Promise<...>) => Promise<...>, { ...; }]' is not assignable to type 'TestFixture<BrowserContext, { authenticatedPage: Page; context: BrowserContext; } & { port: number; storage: LocalStorage; } & PlaywrightTestArgs & PlaywrightTestOptions & PlaywrightWorkerArgs & PlaywrightWorkerOptions>'.
      Type '[({ port }: { authenticatedPage: Page; context: BrowserContext; } & { port: number; storage: LocalStorage; } & PlaywrightTestArgs & PlaywrightTestOptions & PlaywrightWorkerArgs & PlaywrightWorkerOptions, use: (r: BrowserContext) => Promise<...>) => Promise<...>, { ...; }]' provides no match for the signature '(args: { authenticatedPage: Page; context: BrowserContext; } & { port: number; storage: LocalStorage; } & PlaywrightTestArgs & PlaywrightTestOptions & PlaywrightWorkerArgs & PlaywrightWorkerOptions, use: (r: BrowserContext) => Promise<...>, testInfo: TestInfo): any'.ts(2322)
test.d.ts(1151, 3): The expected type comes from property 'context' which is declared here on type 'Fixtures<{ authenticatedPage: Page; context: BrowserContext; }, { port: number; storage: LocalStorage; }, PlaywrightTestArgs & PlaywrightTestOptions, PlaywrightWorkerArgs & PlaywrightWorkerOptions>'
```
</details>

<details><summary>Simplified fixtures showing the error</summary>

```ts
export const lighthouseTest = base.extend<
  {
    context: BrowserContext;
  },
  {
    port: number;
  }
>({
  port: [
    async ({}, use) => {
      const port = await getPort();
      await use(port);
    },
    { scope: 'worker' },
  ],

  // Type conflict here 
  context: [
    async ({ port }, use) => {
      const userDataDir = os.tmpdir();
      const context = await chromium.launchPersistentContext(userDataDir, {
        args: [`--remote-debugging-port=${port}`, '--lang=en-GB'],
      });
      await use(context);
      await context.close();
    },
    { scope: 'test' },
  ],
});
```

</details>

Here's the issue for populating auth state in launchPersistentContext: https://github.com/microsoft/playwright/issues/7634

**Comment by dgozman** [comment]
@badsyntax Thank you for following up. I'll take a look at the TS error, that's unexpected.

**Comment by dgozman** [comment]
I have fixed the types issue in #8139, so overriding `context` fixture with `launchPersistent` should work nicely. Let me close this one, since we have a nice workable solution. Meanwhile, we'll experiment with built-in support for persistent context.

**Comment by lakshbhutani** [comment]
@badsyntax what is the use of 
` lighthouseAuthenticatedPage: [
    async (
      { persistentContext: context, lighthousePage: page, storage },
      use
    ) => {
      // The following uses context.route() to mock some auth related requests (token validation)
      await mockB2CLogin(context);

      await page.goto(rootUrl);
      await insertAuthStorage(page, storage);
      await use(page);
    },
    { scope: 'test' },
  ]`

**Comment by lakshbhutani** [comment]
> ```ts
> await insertAuthStorage(page, storage)
> ```

what is the meaning of  await insertAuthStorage(page, storage)? since you are already injecting auth in  await use(localhostStorage.localStorage); in storage step?
why do we need to do this  await insertAuthStorage(page, storage)

**Comment by lakshbhutani1** [comment]
> Ok i've got this to work with the Test runner.
> 
> Here's my `e2eUtil.ts` file with playwright fixtures:
> 
> ```ts
> import os from 'os';
> import getPort from 'get-port';
> import { playAudit } from 'playwright-lighthouse';
> import { test as base } from '@playwright/test';
> import { BrowserContext, chromium, Page } from 'playwright';
> // @ts-ignore
> import lighthouseMobileConfig from 'lighthouse/lighthouse-core/config/lr-mobile-config.js';
> import { rootUrl } from './constants';
> 
> // auth-state.json is saved during globalSetup
> import authState from './fixtures/auth-state.json';
> 
> type LocalStorage = Array<{ name: string; value: string }>;
> 
> type AuthState = {
>   origins: Array<{
>     origin: string;
>     localStorage: Array<{
>       name: string;
>       value: string;
>     }>;
>   }>;
> };
> 
> export const lighthouseTest = base.extend<
>   {
>     lighthouseAuthenticatedPage: Page;
>     lighthousePage: Page;
>     persistentContext: BrowserContext;
>   },
>   {
>     port: number;
>     storage: LocalStorage;
>   }
> >({
>   // We need to assign a unique port for each playwright worker to allow
>   // lighthouse tests to run in parallel
>   port: [
>     async ({}, use) => {
>       const port = await getPort();
>       await use(port);
>     },
>     { scope: 'worker' },
>   ],
> 
>   // As there's currently no way to insert auth-state into a PersistentContext we
>   // need to explicity load and insert the auth storage. In our case we only care
>   // about localStorage for auth.
>   storage: [
>     async ({}, use) => {
>       const localhostStorage = (authState as AuthState).origins.find(
>         ({ origin }) => {
>           return origin === new URL(rootUrl).origin;
>         }
>       );
>       if (!localhostStorage) {
>         throw new Error(`No storage state found in ${authStatePath}`);
>       }
>       if (!localhostStorage.localStorage.length) {
>         throw new Error(`No localStorage state found in ${authStatePath}`);
>       }
> 
>       await use(localhostStorage.localStorage);
>     },
>     { scope: 'worker' },
>   ],
> 
>   // As lighthouse opens a new page, and as playwright does not by default allow
>   // shared contexts, we need to explicitly create a persistent context to
>   // allow lighthouse to run behind authenticated routes.
>   persistentContext: [
>     async ({ port }, use) => {
>       const userDataDir = os.tmpdir();
>       const context = await chromium.launchPersistentContext(userDataDir, {
>         args: [`--remote-debugging-port=${port}`],
>       });
> 
>       await use(context);
>     },
>     { scope: 'test' },
>   ],
> 
>   lighthousePage: [
>     async ({ persistentContext: context }, use) => {
>       const page = await context.newPage();
>       await use(page);
>     },
>     { scope: 'test' },
>   ],
> 
>   lighthouseAuthenticatedPage: [
>     async (
>       { persistentContext: context, lighthousePage: page, storage },
>       use
>     ) => {
>       // The following uses context.route() to mock some auth related requests (token validation)
>       await mockB2CLogin(context);
> 
>       await page.goto(rootUrl);
>       await insertAuthStorage(page, storage);
>       await use(page);
>     },
>     { scope: 'test' },
>   ],
> });
> 
> type LighthouseThresholds = {
>   performance?: number;
>   accessibility?: number;
>   pwa?: number;
>   'best-practices'?: number;
> };
> 
> export async function lighthouseAudit(
>   page: Page,
>   port: number,
>   thresholds?: LighthouseThresholds,
>   config = lighthouseMobileConfig
> ): Promise<void> {
>   await playAudit({
>     page,
>     config,
>     port,
>     reports: {
>       formats: {
>         html: true,
>         json: true,
>         csv: false,
>       },
>       name: `lhr-${new Date().getTime()}`,
>       directory: './.lighthouse',
>     },
>     thresholds: {
>       performance: 70,
>       accessibility: 95,
>       pwa: 50,
>       'best-practices': 95,
>       ...thresholds,
>     },
>     opts: {
>       logLevel: 'error',
>     },
>   });
> }
> ```
> 
> Running lighthouse on a public route:
> 
> ```ts
> import { rootUrl } from '../../constants';
> import { lighthouseAudit, lighthouseTest } from '../../e2eUtil';
> 
> lighthouseTest.describe('Login screen lighthouse', () => {
>   lighthouseTest(
>     'should pass lighthouse tests',
>     async ({ lighthousePage: page, port }) => {
>       await page.goto(rootUrl);
>       await page.waitForSelector('[data-testid="login-button"]');
>       await lighthouseAudit(page, port);
>     }
>   );
> });
> ```
> 
> Running lighthouse on an authenticated route:
> 
> ```ts
> import { rootUrl } from '../../constants';
> import { lighthouseAudit, lighthouseTest } from '../../e2eUtil';
> 
> lighthouseTest.describe('Dashboard screen lighthouse', () => {
>   lighthouseTest(
>     'should pass lighthouse tests',
>     async ({ port, lighthouseAuthenticatedPage: page }) => {
>       await page.goto(rootUrl);
>       await page.innerText('[data-testid="app-bootstrapped"]');
>       await lighthouseAudit(page, port);
>     }
>   );
> });
> ```
> 
> I'd appreciate a quick review to validate I'm not doing anything wrong with this approach.
> 
> @dgozman It would be convenient if `chromium.launchPersistentContext` provided a way to populate auth storage similar to [browser.newContext({ storageState: 'state.json' })](https://playwright.dev/docs/api/class-browser#browser-new-context)

Can you please attach a working example for this?

**Comment by lakshbhutani** [comment]
> Ok i've got this to work with the Test runner.
> 
> Here's my `e2eUtil.ts` file with playwright fixtures:
> 
> ```ts
> import os from 'os';
> import getPort from 'get-port';
> import { playAudit } from 'playwright-lighthouse';
> import { test as base } from '@playwright/test';
> import { BrowserContext, chromium, Page } from 'playwright';
> // @ts-ignore
> import lighthouseMobileConfig from 'lighthouse/lighthouse-core/config/lr-mobile-config.js';
> import { rootUrl } from './constants';
> 
> // auth-state.json is saved during globalSetup
> import authState from './fixtures/auth-state.json';
> 
> type LocalStorage = Array<{ name: string; value: string }>;
> 
> type AuthState = {
>   origins: Array<{
>     origin: string;
>     localStorage: Array<{
>       name: string;
>       value: string;
>     }>;
>   }>;
> };
> 
> export const lighthouseTest = base.extend<
>   {
>     lighthouseAuthenticatedPage: Page;
>     lighthousePage: Page;
>     persistentContext: BrowserContext;
>   },
>   {
>     port: number;
>     storage: LocalStorage;
>   }
> >({
>   // We need to assign a unique port for each playwright worker to allow
>   // lighthouse tests to run in parallel
>   port: [
>     async ({}, use) => {
>       const port = await getPort();
>       await use(port);
>     },
>     { scope: 'worker' },
>   ],
> 
>   // As there's currently no way to insert auth-state into a PersistentContext we
>   // need to explicity load and insert the auth storage. In our case we only care
>   // about localStorage for auth.
>   storage: [
>     async ({}, use) => {
>       const localhostStorage = (authState as AuthState).origins.find(
>         ({ origin }) => {
>           return origin === new URL(rootUrl).origin;
>         }
>       );
>       if (!localhostStorage) {
>         throw new Error(`No storage state found in ${authStatePath}`);
>       }
>       if (!localhostStorage.localStorage.length) {
>         throw new Error(`No localStorage state found in ${authStatePath}`);
>       }
> 
>       await use(localhostStorage.localStorage);
>     },
>     { scope: 'worker' },
>   ],
> 
>   // As lighthouse opens a new page, and as playwright does not by default allow
>   // shared contexts, we need to explicitly create a persistent context to
>   // allow lighthouse to run behind authenticated routes.
>   persistentContext: [
>     async ({ port }, use) => {
>       const userDataDir = os.tmpdir();
>       const context = await chromium.launchPersistentContext(userDataDir, {
>         args: [`--remote-debugging-port=${port}`],
>       });
> 
>       await use(context);
>     },
>     { scope: 'test' },
>   ],
> 
>   lighthousePage: [
>     async ({ persistentContext: context }, use) => {
>       const page = await context.newPage();
>       await use(page);
>     },
>     { scope: 'test' },
>   ],
> 
>   lighthouseAuthenticatedPage: [
>     async (
>       { persistentContext: context, lighthousePage: page, storage },
>       use
>     ) => {
>       // The following uses context.route() to mock some auth related requests (token validation)
>       await mockB2CLogin(context);
> 
>       await page.goto(rootUrl);
>       await insertAuthStorage(page, storage);
>       await use(page);
>     },
>     { scope: 'test' },
>   ],
> });
> 
> type LighthouseThresholds = {
>   performance?: number;
>   accessibility?: number;
>   pwa?: number;
>   'best-practices'?: number;
> };
> 
> export async function lighthouseAudit(
>   page: Page,
>   port: number,
>   thresholds?: LighthouseThresholds,
>   config = lighthouseMobileConfig
> ): Promise<void> {
>   await playAudit({
>     page,
>     config,
>     port,
>     reports: {
>       formats: {
>         html: true,
>         json: true,
>         csv: false,
>       },
>       name: `lhr-${new Date().getTime()}`,
>       directory: './.lighthouse',
>     },
>     thresholds: {
>       performance: 70,
>       accessibility: 95,
>       pwa: 50,
>       'best-practices': 95,
>       ...thresholds,
>     },
>     opts: {
>       logLevel: 'error',
>     },
>   });
> }
> ```
> 
> Running lighthouse on a public route:
> 
> ```ts
> import { rootUrl } from '../../constants';
> import { lighthouseAudit, lighthouseTest } from '../../e2eUtil';
> 
> lighthouseTest.describe('Login screen lighthouse', () => {
>   lighthouseTest(
>     'should pass lighthouse tests',
>     async ({ lighthousePage: page, port }) => {
>       await page.goto(rootUrl);
>       await page.waitForSelector('[data-testid="login-button"]');
>       await lighthouseAudit(page, port);
>     }
>   );
> });
> ```
> 
> Running lighthouse on an authenticated route:
> 
> ```ts
> import { rootUrl } from '../../constants';
> import { lighthouseAudit, lighthouseTest } from '../../e2eUtil';
> 
> lighthouseTest.describe('Dashboard screen lighthouse', () => {
>   lighthouseTest(
>     'should pass lighthouse tests',
>     async ({ port, lighthouseAuthenticatedPage: page }) => {
>       await page.goto(rootUrl);
>       await page.innerText('[data-testid="app-bootstrapped"]');
>       await lighthouseAudit(page, port);
>     }
>   );
> });
> ```
> 
> I'd appreciate a quick review to validate I'm not doing anything wrong with this approach.
> 
> @dgozman It would be convenient if `chromium.launchPersistentContext` provided a way to populate auth storage similar to [browser.newContext({ storageState: 'state.json' })](https://playwright.dev/docs/api/class-browser#browser-new-context)

Can you please attach a working example for this?

**Comment by rggammon** [comment]
The general idea of insertAuthStorage seems to be to add the playwright auth storage state to the persistent context so that lighthouse can pick it up that way. While the storage gets injected into the playwright context, that doesn't seem to affect lighthouse which is why the persistent context is needed, as I understand it.

eg: the below is working for me as an example.

type AuthState = {
  cookies: Array<{
    name: string;
    value: string;
    domain: string;
    path: string;
    expires: number;
    httpOnly: boolean;
    secure: boolean;
    sameSite: 'Strict' | 'Lax' | 'None';
  }>;
  origins: Array<{
    origin: string;
    localStorage: Array<{
      name: string;
      value: string;
    }>;
  }>;
};

async function insertAuthStorage(page: Page, hostname: string): Promise<void> {
  const authState: AuthState = JSON.parse(
    fs.readFileSync(getAuthArtifactFilePath('state', 'storageState.json'), 'utf8')
  );

  // Set cookies
  for (let cookie of authState.cookies) {
    if (cookie.domain === hostname) {
      const cookieValue = `${cookie.name}=${cookie.value};expires=${new Date(
        cookie.expires * 1000
      ).toUTCString()};path=${cookie.path};domain=${cookie.domain};secure=${cookie.secure};httponly=${
        cookie.httpOnly
      };samesite=${cookie.sameSite}`;

      await page.evaluate((cookieValue) => (document.cookie = cookieValue), cookieValue);
    }
  }

  // Set local storage
  const origin = 'https://' + hostname;
  const localStorageKVPs = authState.origins.find((o) => {
    return o.origin === origin;
  });
  if (localStorageKVPs) {
    for (let kvp of localStorageKVPs.localStorage) {
      await page.evaluate((kvp) => localStorage.setItem(kvp[0], kvp[1]), [kvp.name, kvp.value]);
    }
  }
}

**Comment by CkumarGitHub** [comment]
> Ok i've got this to work with the Test runner.
> 
> Here's my `e2eUtil.ts` file with playwright fixtures:
> 
> ```ts
> import os from 'os';
> import getPort from 'get-port';
> import { playAudit } from 'playwright-lighthouse';
> import { test as base } from '@playwright/test';
> import { BrowserContext, chromium, Page } from 'playwright';
> // @ts-ignore
> import lighthouseMobileConfig from 'lighthouse/lighthouse-core/config/lr-mobile-config.js';
> import { rootUrl } from './constants';
> 
> // auth-state.json is saved during globalSetup
> import authState from './fixtures/auth-state.json';
> 
> type LocalStorage = Array<{ name: string; value: string }>;
> 
> type AuthState = {
>   origins: Array<{
>     origin: string;
>     localStorage: Array<{
>       name: string;
>       value: string;
>     }>;
>   }>;
> };
> 
> export const lighthouseTest = base.extend<
>   {
>     lighthouseAuthenticatedPage: Page;
>     lighthousePage: Page;
>     persistentContext: BrowserContext;
>   },
>   {
>     port: number;
>     storage: LocalStorage;
>   }
> >({
>   // We need to assign a unique port for each playwright worker to allow
>   // lighthouse tests to run in parallel
>   port: [
>     async ({}, use) => {
>       const port = await getPort();
>       await use(port);
>     },
>     { scope: 'worker' },
>   ],
> 
>   // As there's currently no way to insert auth-state into a PersistentContext we
>   // need to explicity load and insert the auth storage. In our case we only care
>   // about localStorage for auth.
>   storage: [
>     async ({}, use) => {
>       const localhostStorage = (authState as AuthState).origins.find(
>         ({ origin }) => {
>           return origin === new URL(rootUrl).origin;
>         }
>       );
>       if (!localhostStorage) {
>         throw new Error(`No storage state found in ${authStatePath}`);
>       }
>       if (!localhostStorage.localStorage.length) {
>         throw new Error(`No localStorage state found in ${authStatePath}`);
>       }
> 
>       await use(localhostStorage.localStorage);
>     },
>     { scope: 'worker' },
>   ],
> 
>   // As lighthouse opens a new page, and as playwright does not by default allow
>   // shared contexts, we need to explicitly create a persistent context to
>   // allow lighthouse to run behind authenticated routes.
>   persistentContext: [
>     async ({ port }, use) => {
>       const userDataDir = os.tmpdir();
>       const context = await chromium.launchPersistentContext(userDataDir, {
>         args: [`--remote-debugging-port=${port}`],
>       });
> 
>       await use(context);
>     },
>     { scope: 'test' },
>   ],
> 
>   lighthousePage: [
>     async ({ persistentContext: context }, use) => {
>       const page = await context.newPage();
>       await use(page);
>     },
>     { scope: 'test' },
>   ],
> 
>   lighthouseAuthenticatedPage: [
>     async (
>       { persistentContext: context, lighthousePage: page, storage },
>       use
>     ) => {
>       // The following uses context.route() to mock some auth related requests (token validation)
>       await mockB2CLogin(context);
> 
>       await page.goto(rootUrl);
>       await insertAuthStorage(page, storage);
>       await use(page);
>     },
>     { scope: 'test' },
>   ],
> });
> 
> type LighthouseThresholds = {
>   performance?: number;
>   accessibility?: number;
>   pwa?: number;
>   'best-practices'?: number;
> };
> 
> export async function lighthouseAudit(
>   page: Page,
>   port: number,
>   thresholds?: LighthouseThresholds,
>   config = lighthouseMobileConfig
> ): Promise<void> {
>   await playAudit({
>     page,
>     config,
>     port,
>     reports: {
>       formats: {
>         html: true,
>         json: true,
>         csv: false,
>       },
>       name: `lhr-${new Date().getTime()}`,
>       directory: './.lighthouse',
>     },
>     thresholds: {
>       performance: 70,
>       accessibility: 95,
>       pwa: 50,
>       'best-practices': 95,
>       ...thresholds,
>     },
>     opts: {
>       logLevel: 'error',
>     },
>   });
> }
> ```
> 
> Running lighthouse on a public route:
> 
> ```ts
> import { rootUrl } from '../../constants';
> import { lighthouseAudit, lighthouseTest } from '../../e2eUtil';
> 
> lighthouseTest.describe('Login screen lighthouse', () => {
>   lighthouseTest(
>     'should pass lighthouse tests',
>     async ({ lighthousePage: page, port }) => {
>       await page.goto(rootUrl);
>       await page.waitForSelector('[data-testid="login-button"]');
>       await lighthouseAudit(page, port);
>     }
>   );
> });
> ```
> 
> Running lighthouse on an authenticated route:
> 
> ```ts
> import { rootUrl } from '../../constants';
> import { lighthouseAudit, lighthouseTest } from '../../e2eUtil';
> 
> lighthouseTest.describe('Dashboard screen lighthouse', () => {
>   lighthouseTest(
>     'should pass lighthouse tests',
>     async ({ port, lighthouseAuthenticatedPage: page }) => {
>       await page.goto(rootUrl);
>       await page.innerText('[data-testid="app-bootstrapped"]');
>       await lighthouseAudit(page, port);
>     }
>   );
> });
> ```
> 
> I'd appreciate a quick review to validate I'm not doing anything wrong with this approach.
> 
> @dgozman It would be convenient if `chromium.launchPersistentContext` provided a way to populate auth storage similar to [browser.newContext({ storageState: 'state.json' })](https://playwright.dev/docs/api/class-browser#browser-new-context)

@badsyntax :
I am getting below error for authenticated + global setup
TypeError: Failed to fetch browser webSocket URL from http://127.0.0.1:60803/json/version: fetch failed
If I just include page fixture and then use page.goto(url), then it opens up the browser (but still the auth state is not loaded)
